### ISSUE TITLE
Non Empty Vector

### ISSUE THREAD

**Issue by Firstyear** [kanidm/kanidm#267]
In a lot of places we use a vector, but we have to assume 0 to infinite items. We need a non-empty vec type, probably based on small vec for storage of entry attributes (away from btreeset to improve allocations)

**Comment by MinhPhan8803** [comment]
Is this requirement only concerning parse types (similar to #1452 ) or do we have other use cases where it is desired?
Also how should this vec behave when size reaches 0?

**Comment by Firstyear** [comment]
it should never let you reach 0, by refusing the last "remove" operation from proceeding. 

It's not just for parsers, but some internal types it would be useful for us to be able to prevent and detect cases where structures that must have at least one value.

It would also be useful in our proto structures so that clients can see that values always have at least one. 

It was a while ago that I looked into libraries that provided non-empty vecs but they all had some limitation at some point in the past. However that may have changed now. 

A good way to think about it is numbers 0, 1 and infinite. We have Option<T> which is 0 or 1. We have Vec which is 0 to infinite. But we don't have NonEmptyVec for 1 to infinite.

**Comment by MinhPhan8803** [comment]
Is there some specific example of the requirements our non empty vec that existing crates (in the past) couldn't meet?
Also if I'm not mistaken, refusing the "remove" operation could mean something like returning an `Err()` right? (Or panicking but that'd be awful).

**Comment by MinhPhan8803** [comment]
I'm thinking of testing with this crate first https://docs.rs/nonempty/0.8.1/nonempty/.
It has a Vec like API and serde support. The obscure name is a problem though.

**Comment by Firstyear** [comment]
I don't mind the name? looks like a good crate though, so I'd be keen to use it as a solution to this issue.

**Comment by MinhPhan8803** [comment]
Aight I'm gonna try the crate out.

**Comment by MinhPhan8803** [comment]
Would `DbEntryV1` in this file https://github.com/kanidm/kanidm/blob/master/server/lib/src/be/dbentry.rs be candidate for the the non empty vec? Since the implementation is essentially "this should never happen".

**Comment by Firstyear** [comment]
Yep it would, but we'd need to handle that error properly up through the be code :) 

With a change like this we don't have to do everything at once, we can just do a bit at a time btw.

**Comment by MinhPhan8803** [comment]
Yeah I will probably try to handle this first. Currently I'm just going through all the Vec's in code, see what they look like.

**Comment by MinhPhan8803** [comment]
The password quality feedback https://github.com/kanidm/kanidm/blob/master/server/lib/src/idm/credupdatesession.rs and `CredState::Continue` https://github.com/kanidm/kanidm/blob/master/server/lib/src/idm/authsession.rs look like other candidates for the non empty vec. Both because having them be empty don't make a whole lot of sense, and the password feedback has the `None` case handled separately anyway.

**Comment by MinhPhan8803** [comment]
Actually the feedback can be empty from the zxcvbn doc so probably not that.

**Comment by MinhPhan8803** [comment]
A lot of places that enforce non empty actually use a btreeset, could we try to change some to non empty?
Also some of the enum variants defined in proto v1 I feel could use non empty, though I'm not sure how valid that would be.

**Comment by Firstyear** [comment]
I think that is comes back to what I said previously about looking for structures that define 1 to N items and rely on .is_empty() to error during their construction. I think it's also important to remember that btreeset is probably being used to make searching/lookups faster on those structures too. So it's very hard to just blanket define if something is or is not valid for this. Having nonEmpty now available means that this can be applied slowly and over time too and it's available for new code also. 

So I'm not sure specifically what else "needs nonempty" right now, but I know there will be things.

**Comment by MinhPhan8803** [comment]
I see. Thanks for clarifying!

**Comment by Firstyear** [comment]
It's one of the challenges of working on "someone elses project" - I have a really large mental map of all the code and the in-and-outs and what each part does that takes time for someone new like yourself to learn and get to. It does get easier over time though :)

**Comment by yaleman** [comment]
#1522, #1501, #1543 refer to this
